### PR TITLE
Add Network Policy IPBlock support

### PR DESCRIPTION
**Pull request by tong101**
K8s network policy supports both ingress and egress policy when
peer is from a IPBlock CIDR. The traffic from IPBlock except will
be denied. This patch adds support for this feature.

Signed-off-by: Tong Liu <tongl@vmware.com>

### CODE REVIEW

**Comment by tong101** [comment]
ready for review

**Comment by rajatchopra** [comment]
Looks good to me. Can you think of an example NetworkPolicy and set of pods/namespaces that could go about to test this scenario?

**Comment by tong101** [comment]
@rajatchopra Here is an example:
https://github.com/tong101/ovn-kubeconfig/blob/master/policy/policy2.yaml

**Comment by shettyg** [comment]
I see 2 issues.
1. Looks like if you have a ipblock section, the code will simply not allow any l4 ports in matches.
2. It will also skip rules for namespaces and pod selectors for that l4 ports.

**Comment by tong101** [comment]
@shettyg fixed the two issues. Could you please revisit it?

**Comment by shettyg** [comment]
@tong101  I see another issue. If you delete the policy, the deny rule added for ipblock does not get deleted. This is because deleteAclsPolicy() does not catch the ipblock deny rule. This is because addACLDeny does not take policy as an input.

**Comment by tong101** [comment]
@shettyg It gets ugly to consolidate into one func so that I separate the ACL func for IPBlock. Fixed the issue in latest ps.

File: go-controller/pkg/ovn/policy.go [hunk 1]
```diff
@@ -222,8 +243,100 @@ func (oc *Controller) deleteACLAllow(namespace, policy, logicalSwitch,
...
+	} else {
+		portMatch = fmt.Sprintf("inport == \\\"%s\\\"", logicalPort)
+		l3Match = fmt.Sprintf("ip4.dst == %s", except)
+		match = fmt.Sprintf("match=\"%s && %s\"", portMatch, l3Match)
+		direction = fromLport
+	}
+
+	uuid, err := exec.Command(OvnNbctl, "--data=bare", "--no-heading",
+		"--columns=_uuid", "find", "ACL", match, "action=drop",
+		fmt.Sprintf("external-ids:default-deny-direction=%s", direction),
```
Comment (shettyg) [inline go-controller/pkg/ovn/policy.go]: We should change this name from "default-deny-direction" to ipblock-deny-direction. Same in couple of other places.

File: go-controller/pkg/ovn/policy.go [hunk 2]
```diff
@@ -414,33 +527,85 @@ func getL3MatchFromAddressSet(addressSets []string, policyType string) string {
 	return l3Match
 }
 
+func getMatchFromIPBlock(ipBlockCidr, portMatch, l4Match,
```
Comment (shettyg) [inline go-controller/pkg/ovn/policy.go]: Would you mind changing "portMatch" to "lportMatch".  When I read, portMatch can mean tcp and udp ports and that causes confusions.

File: go-controller/pkg/ovn/policy.go (hunk 1 above)
Comment (tong101) [reply go-controller/pkg/ovn/policy.go]: Good point, updated to ipblock-deny-direction.

File: go-controller/pkg/ovn/policy.go (hunk 2 above)
Comment (tong101) [reply go-controller/pkg/ovn/policy.go]: Done

File: go-controller/pkg/ovn/policy.go [hunk 3]
```diff
@@ -222,8 +243,100 @@ func (oc *Controller) deleteACLAllow(namespace, policy, logicalSwitch,
 	}
 }
 
-func (oc *Controller) addACLDeny(namespace, logicalSwitch,
-	logicalPort, policyType string) {
+func (oc *Controller) addIPBlockACLDeny(namespace, policy, logicalSwitch,
+	logicalPort, policyType, except, priority string) {
+	var match, l3Match, direction, portMatch string
```
Comment (shettyg) [inline go-controller/pkg/ovn/policy.go]: Would you mind changing portMatch to lportMatch at these places too?

Comment (tong101) [reply go-controller/pkg/ovn/policy.go]: Done

**Pull request merged**